app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryViewModel.kt

```diff
@@ -0,0 +1,92 @@
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
+import io.reactivex.schedulers.Schedulers
+import io.zenandroid.onlinego.data.model.local.VisitedPuzzleCollection
+import io.zenandroid.onlinego.data.model.local.PuzzleCollection
+import io.zenandroid.onlinego.data.ogs.OGSRestService
+import io.zenandroid.onlinego.data.repositories.PuzzleRepository
+import io.zenandroid.onlinego.mvi.MviView
+import io.zenandroid.onlinego.mvi.Store
+import io.zenandroid.onlinego.utils.addToDisposable
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.update
+import org.koin.core.context.GlobalContext
+
+class PuzzleDirectoryViewModel (
+    private val puzzleRepository: PuzzleRepository,
+    private val restService: OGSRestService,
+    private val store: Store<PuzzleDirectoryState, PuzzleDirectoryAction>
+): ViewModel() {
+    private val _state = MutableStateFlow(PuzzleDirectoryState())
+    val state: StateFlow<PuzzleDirectoryState> = _state
+    private val subscriptions = CompositeDisposable()
+
+    private val wiring = store.wire()
+    private var viewBinding: Disposable? = null
+
+    override fun onCleared() {
+        wiring.dispose()
+    }
+
+    init {
+        puzzleRepository.getAllPuzzleCollections()
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(this::addCollections, this::onError)
+            .addToDisposable(subscriptions)
+        puzzleRepository.getRecentPuzzleCollections()
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(this::setRecentCollections, this::onError)
+            .addToDisposable(subscriptions)
+        puzzleRepository.getPuzzleCollectionSolutions()
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(this::setCollectionSolutions, this::onError)
+            .addToDisposable(subscriptions)
+    }
+
+    private fun addCollections(nextCollections: List<PuzzleCollection>) {
+        _state.update {
+            it.copy(
+                collections = it.collections.plus(nextCollections.associateBy(PuzzleCollection::id))
+            )
+        }
+    }
+
+    private fun setRecentCollections(recents: List<VisitedPuzzleCollection>) {
+        _state.update {
+            it.copy(
+                recents = recents.associateBy(VisitedPuzzleCollection::timestamp).plus(it.recents)
+            )
+        }
+    }
+
+    private fun setCollectionSolutions(solutions: Map<Long, Int>) {
+        _state.update {
+            it.copy(solutions = solutions)
+        }
+    }
+
+    fun bind(view: MviView<PuzzleDirectoryState, PuzzleDirectoryAction>) {
+        viewBinding = store.bind(view)
+    }
+
+    fun unbind() {
+        viewBinding?.dispose()
+    }
+
+    private fun onError(t: Throwable) {
```

Comment:
`recordException(error)` please

Reply:
I also removed the toasts, I think I was mainly using them for debugging and they slightly violate the paradigm